handlers/jiochat/jiochat.go

```diff
@@ -0,0 +1,323 @@
+	err := handlers.DecodeAndValidateQueryParams(jcVerify, r)
+	if err != nil {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, err)
+	}
+
+	dictOrder := []string{channel.StringConfigForKey(configJiochatAppSecret, ""), jcVerify.Timestamp, jcVerify.Nonce}
+	sort.Sort(sort.StringSlice(dictOrder))
+
+	combinedParams := strings.Join(dictOrder, "")
+
+	hash := sha1.New()
+	hash.Write([]byte(combinedParams))
+	encoded := hex.EncodeToString(hash.Sum(nil))
+
+	ResponseText := "unknown request"
+	StatusCode := 400
+
+	if encoded == jcVerify.Signature {
+		ResponseText = jcVerify.EchoStr
+		StatusCode = 200
+		go h.doFetchAccessToken(channel)
+	}
+
+	w.Header().Set("Content-Type", "text/plain")
+	w.WriteHeader(StatusCode)
+	_, err = fmt.Fprint(w, ResponseText)
+	return nil, err
+}
+
+type moMsg struct {
+	FromUsername string `json:"FromUserName"    validate:"required"`
+	MsgType      string `json:"MsgType"         validate:"required"`
+	CreateTime   int64  `json:"CreateTime"`
+	MsgID        int64  `json:"MsgId"`
+	Event        string `json:"Event"`
+	Content      string `json:"Content"`
+	MediaID      string `json:"MediaId"`
+}
+
+// ReceiveMessage is our HTTP handler function for incoming messages
+func (h *handler) ReceiveMessage(ctx context.Context, channel courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.Event, error) {
+	jcRequest := &moMsg{}
+	err := handlers.DecodeAndValidateJSON(jcRequest, r)
+	if err != nil {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, err)
+	}
+
+	if jcRequest.MsgID == 0 && jcRequest.Event == "" {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, fmt.Errorf("missing parameters, must have either 'MsgId' or 'Event'"))
+	}
+
+	date := time.Unix(jcRequest.CreateTime, 0).UTC()
+
+	urn := urns.NewURNFromParts(urns.JiochatScheme, jcRequest.FromUsername, "")
+
+	if jcRequest.MsgType == "event" && jcRequest.Event == "subscribe" {
+
+		// build the channel event
+		channelEvent := h.Backend().NewChannelEvent(channel, courier.NewConversation, urn)
+
+		err := h.Backend().WriteChannelEvent(ctx, channelEvent)
+		if err != nil {
+			return nil, err
+		}
+
+		return []courier.Event{channelEvent}, courier.WriteChannelEventSuccess(ctx, w, r, channelEvent)
+	}
+
+	if jcRequest.MsgType == "event" {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, fmt.Errorf("unknown event"))
+	}
+
+	// create our message
+	msg := h.Backend().NewIncomingMsg(channel, urn, jcRequest.Content).WithExternalID(fmt.Sprintf("%d", jcRequest.MsgID)).WithReceivedOn(date)
+	if jcRequest.MsgType == "image" || jcRequest.MsgType == "video" || jcRequest.MsgType == "voice" {
+		mediaURL := resolveMediaID(jcRequest.MediaID)
+		msg.WithAttachment(mediaURL)
+	}
+
+	err = h.Backend().WriteMsg(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	return []courier.Event{msg}, courier.WriteMsgSuccess(ctx, w, r, []courier.Msg{msg})
+}
+
+func resolveMediaID(mediaID string) string {
```

Comment:
This isn't really resolving, just building a media URL, so maybe something more like `buildMediaURL` would fit better.

Reply:
okay